File: mbw/src/main/java/com/mycelium/wallet/activity/modern/adapter/AccountListAdapter.java
package com.mycelium.wallet.activity.modern.adapter;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.SharedPreferences;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mycelium.wallet.MbwManager;
import com.mycelium.wallet.R;
import com.mycelium.wallet.activity.modern.RecordRowBuilder;
import com.mycelium.wallet.activity.modern.adapter.holder.AccountViewHolder;
import com.mycelium.wallet.activity.modern.adapter.holder.ArchivedGroupTitleViewHolder;
import com.mycelium.wallet.activity.modern.adapter.holder.GroupTitleViewHolder;
import com.mycelium.wallet.activity.modern.adapter.holder.TotalViewHolder;
import com.mycelium.wallet.activity.modern.model.accounts.AccountItem;
import com.mycelium.wallet.activity.modern.model.accounts.AccountsListModel;
import com.mycelium.wallet.activity.modern.model.ViewAccountModel;
import com.mycelium.wallet.event.AccountGroupCollapsed;
import com.mycelium.wapi.wallet.WalletAccount;
import com.mycelium.wapi.wallet.currency.CurrencySum;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class AccountListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    public static final int GROUP_TITLE_TYPE = 2;
    public static final int ACCOUNT_TYPE = 3;
    public static final int TOTAL_BALANCE_TYPE = 4;
    public static final int GROUP_ARCHIVED_TITLE_TYPE = 5;

    private List<AccountItem> itemList = new ArrayList<>();
    private UUID focusedAccountId;

    private ItemClickListener itemClickListener;
    private ItemSelectListener itemSelectListener;

    private MbwManager mbwManager;
    private RecordRowBuilder builder;
    private LayoutInflater layoutInflater;
    private SharedPreferences pagePrefs;

    public AccountListAdapter(Fragment fragment, MbwManager mbwManager) {
        AccountsListModel listModel = ViewModelProviders.of(fragment).get(AccountsListModel.class);
        Context context = fragment.getContext();
        this.mbwManager = mbwManager;

        layoutInflater = LayoutInflater.from(context);
        builder = new RecordRowBuilder(mbwManager, context.getResources());
        pagePrefs = context.getSharedPreferences("account_list", Context.MODE_PRIVATE);
        listModel.getAccountsData().observe(fragment, new Observer<List<? extends AccountItem>>() {
            @Override
            public void onChanged(List<? extends AccountItem> accountItems) {
                itemList.clear();
                itemList.addAll(accountItems);
                notifyDataSetChanged();
            }
        });
        itemList.addAll(listModel.getAccountsData().getValue());
    }

    public void setItemClickListener(ItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    public void setItemSelectListener(ItemSelectListener itemSelectListener) {
        this.itemSelectListener = itemSelectListener;
    }

    public WalletAccount getFocusedAccount() {
        return mbwManager.getWalletManager(false).getAccount(focusedAccountId);
    }

    public void setFocusedAccountId(UUID focusedAccountId) {
        this.focusedAccountId = focusedAccountId;
        notifyDataSetChanged(); // Somewhy in other cases android would not update alpha value
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder result = null;
        if (viewType == GROUP_TITLE_TYPE) {
            View view = layoutInflater.inflate(R.layout.accounts_title_view, parent, false);
            GroupTitleViewHolder res = new GroupTitleViewHolder(view);
            res.tvBalance.setEventBus(mbwManager.getEventBus());
            res.tvBalance.setCurrencySwitcher(mbwManager.getCurrencySwitcher());
            result = res;
        } else if (viewType == GROUP_ARCHIVED_TITLE_TYPE) {
            View view = layoutInflater.inflate(R.layout.accounts_archived_title_view, parent, false);
            result = new ArchivedGroupTitleViewHolder(view);
        } else if (viewType == ACCOUNT_TYPE) {
            View view = layoutInflater.inflate(R.layout.record_row, parent, false);
            result = new AccountViewHolder(view);
        } else if (viewType == TOTAL_BALANCE_TYPE) {
            View view = layoutInflater.inflate(R.layout.record_row_total, parent, false);
            TotalViewHolder res = new TotalViewHolder(view);
            res.tcdBalance.setCurrencySwitcher(mbwManager.getCurrencySwitcher());
            res.tcdBalance.setEventBus(mbwManager.getEventBus());
            result = res;
        }
        return result;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        final AccountItem item = itemList.get(position);
        int viewType = item.getType();
        if (viewType == ACCOUNT_TYPE) {
            AccountViewHolder accountHolder = (AccountViewHolder) holder;
            final ViewAccountModel account = item.getWalletAccount();
            builder.buildRecordView(accountHolder, account
                    , Objects.equals(mbwManager.getSelectedAccount().getId(), account.accountId)
                    , Objects.equals(focusedAccountId, account.accountId));
            accountHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    setFocusedAccountId(account.accountId);
                    if (itemSelectListener != null) {
                        itemSelectListener.onClick(mbwManager.getWalletManager(false)
                                .getAccount(account.accountId));
                    }

                }
            });
            accountHolder.llAddress.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    setFocusedAccountId(account.accountId);
                    if (itemClickListener != null) {
                        itemClickListener.onItemClick(mbwManager.getWalletManager(false)
                                .getAccount(account.accountId));
                    }
                }
            });
        } else if (viewType == GROUP_TITLE_TYPE) {
            GroupTitleViewHolder groupHolder = (GroupTitleViewHolder) holder;
            groupHolder.tvTitle.setText(Html.fromHtml(item.getTitle()));
            int count = item.getWalletAccountList().size();
            groupHolder.tvAccountsCount.setVisibility(count > 0 ? View.VISIBLE : View.GONE);
            groupHolder.tvAccountsCount.setText("(" + count + ")");
            groupHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    boolean isGroupVisible = !pagePrefs.getBoolean(item.getTitle(), true);
                    pagePrefs.edit().putBoolean(item.getTitle(), isGroupVisible).apply();
                    mbwManager.getEventBus().post(new AccountGroupCollapsed());
                }
            });
            groupHolder.expandIcon.setRotation(pagePrefs.getBoolean(item.getTitle(), true) ? 180 : 0);
            CurrencySum sum = getSpendableBalance(item.getWalletAccountList());
            groupHolder.tvBalance.setValue(sum);
            groupHolder.tvBalance.setVisibility(View.VISIBLE);
        } else if (viewType == GROUP_ARCHIVED_TITLE_TYPE) {
            ArchivedGroupTitleViewHolder groupHolder = (ArchivedGroupTitleViewHolder) holder;
            groupHolder.tvTitle.setText(Html.fromHtml(item.getTitle()));
            int count = item.getWalletAccountList().size();
            groupHolder.tvAccountsCount.setVisibility(count > 0 ? View.VISIBLE : View.GONE);
            groupHolder.tvAccountsCount.setText("(" + count + ")");
            groupHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    boolean isGroupVisible = !pagePrefs.getBoolean(item.getTitle(), true);
                    pagePrefs.edit().putBoolean(item.getTitle(), isGroupVisible).apply();
                    mbwManager.getEventBus().post(new AccountGroupCollapsed());
                }
            });
            groupHolder.expandIcon.setRotation(pagePrefs.getBoolean(item.getTitle(), true) ? 180 : 0);
        } else if (viewType == TOTAL_BALANCE_TYPE) {
            TotalViewHolder totalHolder = (TotalViewHolder) holder;
            CurrencySum sum = getSpendableBalance(item.getWalletAccountList());
            totalHolder.tcdBalance.setValue(sum);
        }
    }

    private CurrencySum getSpendableBalance(List<ViewAccountModel> walletAccountList) {
        CurrencySum currencySum = new CurrencySum();
        for (ViewAccountModel account : walletAccountList) {
            if (account.isActive) {
                currencySum.add(account.balance.confirmed);
            }
        }
        return currencySum;
    }

    @Override
    public int getItemCount() {
        return itemList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return itemList.get(position).getType();
    }

    public interface ItemClickListener {
        void onItemClick(WalletAccount account);
    }

    public interface ItemSelectListener {
        void onClick(WalletAccount account);
    }
}

File: mbw/src/main/java/com/mycelium/wallet/activity/modern/adapter/AccountListAdapter.kt
package com.mycelium.wallet.activity.modern.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.text.Html
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.mycelium.bequant.BequantPreference
import com.mycelium.bequant.common.equalsValuesBy
import com.mycelium.bequant.remote.repositories.Api
import com.mycelium.wallet.MbwManager
import com.mycelium.wallet.R
import com.mycelium.wallet.activity.modern.RecordRowBuilder
import com.mycelium.wallet.activity.modern.adapter.holder.*
import com.mycelium.wallet.activity.modern.model.ViewAccountModel
import com.mycelium.wallet.activity.modern.model.accounts.*
import com.mycelium.wallet.activity.modern.model.accounts.AccountListItem.Type.*
import com.mycelium.wallet.exchange.ValueSum
import com.mycelium.wapi.wallet.Address
import com.mycelium.wapi.wallet.WalletAccount
import java.util.*

class AccountListAdapter(fragment: Fragment, private val mbwManager: MbwManager)
    : ListAdapter<AccountListItem, RecyclerView.ViewHolder>(ItemListDiffCallback()) {
    private val context = fragment.requireContext()

    private var focusedAccountId: UUID? = null
    private var selectedAccountId: UUID? = mbwManager.selectedAccount.id

    private var itemClickListener: ItemClickListener? = null
    var investmentAccountClickListener: ItemClickListener? = null
    private val layoutInflater = LayoutInflater.from(context)
    private val pagePrefs = context.getSharedPreferences("account_list", Context.MODE_PRIVATE)
    private val listModel by fragment.activityViewModels<AccountsListModel>()
    private val walletManager = mbwManager.getWalletManager(false)

    val focusedAccount: WalletAccount<out Address>?
        get() = (focusedAccountId?.let {
            Log.e("!!!!", "focusedAccountId=$it")
            walletManager.getAccount(it)
        }).apply {
            Log.e("!!!!", "focusedAccountId1=$this")
        }

    init {
        listModel.accountsData.observe(fragment, Observer { accountsGroupModels ->
            accountsGroupModels!!
            val selectedAccountExists = accountsGroupModels.any { it.accountsList.any { it is AccountViewModel && it.accountId == selectedAccountId } }
            if (!selectedAccountExists) {
                setFocusedAccountId(null)
            }
            refreshList(accountsGroupModels)
        })
        val accountsGroupsList = listModel.accountsData.value!!
        refreshList(accountsGroupsList)

        Api.accountRepository.accountBalanceGet(fragment.lifecycleScope, {
            val find = it?.find { it.currency?.toLowerCase() == "btc" }
        }, { _, _ ->

        }, {

        })
    }

    private fun refreshList(accountsGroupModels: List<AccountsGroupModel>) {
        submitList(generateListView(accountsGroupModels))
    }

    private fun generateListView(accountsGroupsList: List<AccountsGroupModel>): List<AccountListItem> {
        val itemList = ArrayList<AccountListItem>()
        var totalAdded = false

        for (accountsGroup in accountsGroupsList) {
            if (accountsGroup.getType() == GROUP_ARCHIVED_TITLE_TYPE) {
                itemList.add(TotalViewModel(getSpendableBalance(itemList)))
                totalAdded = true
            }
            val groupModel = AccountsGroupModel(accountsGroup)
            itemList.add(groupModel)
            groupModel.isCollapsed = !pagePrefs.getBoolean(accountsGroup.getTitle(context), true)
            if (!groupModel.isCollapsed) {
                itemList.addAll(accountsGroup.accountsList)
            }
        }
        if (itemList.isNotEmpty() && !totalAdded) {
            itemList.add(TotalViewModel(getSpendableBalance(itemList)))
        }

        return itemList
    }

    fun setItemClickListener(itemClickListener: ItemClickListener) {
        this.itemClickListener = itemClickListener
    }

    fun setFocusedAccountId(focusedAccountId: UUID?) {
        Log.e("!!!!", "setFocusedAccountId=$focusedAccountId")
        if (this.focusedAccountId == null) {
            this.focusedAccountId = mbwManager.selectedAccount.id
        }
        val oldFocusedPosition = findPosition(this.focusedAccountId)
        // If old account was removed we don't want to notify removed element. It would be updated itself.
        val updateOld = walletManager.getAccount(this.focusedAccountId!!) != null
        val oldSelectedPosition = findPosition(this.selectedAccountId)
        this.focusedAccountId = focusedAccountId
        if (focusedAccountId != null && walletManager.getAccount(focusedAccountId)?.isActive == true) {
            this.selectedAccountId = focusedAccountId
            notifyItemChanged(oldSelectedPosition)
        }
        if (updateOld) {
            notifyItemChanged(oldFocusedPosition)
        }
        notifyItemChanged(findPosition(this.focusedAccountId))
    }

    private fun findPosition(account: UUID?): Int {
        var position = -1
        for (i in 0 until itemCount) {
            if (getItem(i).getType() == ACCOUNT_TYPE) {
                val item = getItem(i) as AccountViewModel
                if (item.accountId == account) {
                    position = i
                    break
                }
            }
        }
        return position
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (AccountListItem.Type.fromId(viewType)) {
            GROUP_TITLE_TYPE -> createGroupViewHolder(parent)
            GROUP_ARCHIVED_TITLE_TYPE -> createArchivedTitleViewHolder(parent)
            ACCOUNT_TYPE -> createAccountViewHolder(parent)
            TOTAL_BALANCE_TYPE -> createTotalBalanceViewHolder(parent)
            INVESTMENT_TYPE -> createInvestmentAccountViewHolder(parent)
            else -> throw IllegalArgumentException("Unknown account type")
        }
    }

    private fun createGroupViewHolder(parent: ViewGroup): GroupTitleViewHolder {
        val view = layoutInflater.inflate(R.layout.accounts_title_view, parent, false)
        return GroupTitleViewHolder(view)
    }

    private fun createArchivedTitleViewHolder(parent: ViewGroup): ArchivedGroupTitleViewHolder {
        val view = layoutInflater.inflate(R.layout.archived_accounts_title_view, parent, false)
        return ArchivedGroupTitleViewHolder(view)
    }

    private fun createAccountViewHolder(parent: ViewGroup): AccountViewHolder {
        val view = layoutInflater.inflate(R.layout.record_row, parent, false)
        return AccountViewHolder(view)
    }

    private fun createTotalBalanceViewHolder(parent: ViewGroup): TotalViewHolder {
        val view = layoutInflater.inflate(R.layout.record_row_total, parent, false)
        return TotalViewHolder(view)
    }

    private fun createInvestmentAccountViewHolder(parent: ViewGroup): InvestmentViewHolder =
            InvestmentViewHolder(layoutInflater.inflate(R.layout.record_row_investment, parent, false))

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = getItem(position)
        val viewType = getItemViewType(position)
        when (AccountListItem.Type.fromId(viewType)) {
            ACCOUNT_TYPE -> {
                val accountHolder = holder as AccountViewHolder
                val account = item as AccountViewModel
                val viewModel = ViewAccountModel(account, context)
                val builder = RecordRowBuilder(mbwManager, context.resources)
                builder.buildRecordView(accountHolder, viewModel, mbwManager.selectedAccount.id == account.accountId,
                        focusedAccountId == account.accountId)
                accountHolder.llAddress.setOnClickListener {
                    setFocusedAccountId(account.accountId)
                    walletManager.getAccount(account.accountId)?.run { itemClickListener?.onItemClick(this) }
                }
            }
            GROUP_TITLE_TYPE -> {
                val groupHolder = holder as GroupTitleViewHolder
                val group = item as AccountsGroupModel
                buildGroupBase(group, groupHolder)
                groupHolder.tvBalance?.coinType = group.coinType
                groupHolder.tvBalance?.setValue(group.sum!!, false)
                groupHolder.tvBalance?.visibility = View.VISIBLE
            }
            GROUP_ARCHIVED_TITLE_TYPE -> {
                val groupHolder = holder as ArchivedGroupTitleViewHolder
                val group = item as AccountsGroupModel
                buildGroupBase(group, groupHolder)
            }
            TOTAL_BALANCE_TYPE -> {
                val totalHolder = holder as TotalViewHolder
                val sum = (item as TotalViewModel).balance
                totalHolder.tcdBalance.setValue(sum, totalBalance = true)
            }
            INVESTMENT_TYPE -> {
                val investHolder = holder as InvestmentViewHolder
                val investItem = item as AccountInvestmentViewModel
                investHolder.balance.text = investItem.balance
                investHolder.itemView.setOnClickListener {
                    investmentAccountClickListener?.onItemClick(item.account)
                }
                if (BequantPreference.isLogged()) {
                    investHolder.balance.visibility = View.VISIBLE
                    investHolder.activateLink.visibility = View.GONE
                } else {
                    investHolder.balance.visibility = View.GONE
                    investHolder.activateLink.visibility = View.VISIBLE
                }
            }
            UNKNOWN -> throw IllegalArgumentException("Unknown view type")
            else -> {}
        }
    }

    @SuppressLint("SetTextI18n")
    private fun buildGroupBase(group: AccountsGroupModel, groupHolder: GroupTitleViewHolder) {
        val title = group.getTitle(context)
        groupHolder.tvTitle.text = Html.fromHtml(title)
        val count = group.accountsList.size
        groupHolder.tvAccountsCount.visibility = if (count > 0 && !group.isInvestmentAccount) View.VISIBLE else View.GONE
        groupHolder.tvAccountsCount.text = "($count)"
        groupHolder.itemView.setOnClickListener {
            //Should be here as initial state in model is wrong
            val isGroupVisible = !pagePrefs.getBoolean(title, true)
            pagePrefs.edit().putBoolean(title, isGroupVisible).apply()
            refreshList(listModel.accountsData.value!!)
        }
        groupHolder.expandIcon.rotation = (if (!group.isCollapsed) 180 else 0).toFloat()
    }

    private fun getSpendableBalance(walletAccountList: List<AccountListItem>): ValueSum {
        val sum = ValueSum()
        for (item in walletAccountList) {
            if (item.getType() == GROUP_TITLE_TYPE) {
                for (account in (item as AccountsGroupModel).accountsList) {
                    if (account.getType() == ACCOUNT_TYPE && (account as AccountViewModel).isActive) {
                        sum.add(account.balance!!.spendable)
                    }
                    if (account.getType() == INVESTMENT_TYPE) {
                        val account = account as AccountInvestmentViewModel
                        sum.add(account.account.accountBalance.spendable)
                    }
                }
            }
        }
        return sum
    }

    override fun getItemViewType(position: Int) = getItem(position).getType().typeId

    interface ItemClickListener {
        fun onItemClick(account: WalletAccount<out Address>)
    }

    open class ItemListDiffCallback() : DiffUtil.ItemCallback<AccountListItem>() {
        override fun areItemsTheSame(oldItem: AccountListItem, newItem: AccountListItem): Boolean =
             when {
                oldItem.getType() != newItem.getType() -> false
                listOf(GROUP_TITLE_TYPE, GROUP_ARCHIVED_TITLE_TYPE).any { it == oldItem.getType() } -> {
                    (oldItem as AccountsGroupModel).titleId == (newItem as AccountsGroupModel).titleId
                }
                oldItem.getType() == ACCOUNT_TYPE -> {
                    (oldItem as AccountViewModel).accountId == (newItem as AccountViewModel).accountId
                }
                oldItem.getType() == INVESTMENT_TYPE -> {
                    (oldItem as AccountInvestmentViewModel).accountId == (newItem as AccountInvestmentViewModel).accountId
                }
                else -> true
            }

        override fun areContentsTheSame(oldItem: AccountListItem, newItem: AccountListItem): Boolean =
                when (oldItem.getType()) {
                    GROUP_TITLE_TYPE, GROUP_ARCHIVED_TITLE_TYPE -> {
                        equalsValuesBy(newItem as AccountsGroupModel, oldItem as AccountsGroupModel,
                                { it.isCollapsed }, { it.coinType }, { it.accountsList.size },
                                { it.sum })
                    }
                    ACCOUNT_TYPE -> {
                        equalsValuesBy(newItem as AccountViewModel, oldItem as AccountViewModel,
                                { it.displayAddress }, { it.isActive }, { it.canSpend },
                                { it.externalAccountType }, { it.isRMCLinkedAccount }, { it.label },
                                { it.showBackupMissingWarning }, { it.syncTotalRetrievedTransactions },
                                { it.isSyncing }, { it.privateKeyCount }, { it.balance?.spendable },
                                { it.isSyncError }, { it.isSelected })
                    }
                    TOTAL_BALANCE_TYPE -> {
                        equalsValuesBy(newItem as TotalViewModel, oldItem as TotalViewModel,
                                {it.balance.values})
                    }
                    INVESTMENT_TYPE -> {
                        equalsValuesBy(newItem as AccountInvestmentViewModel, oldItem as AccountInvestmentViewModel,
                                { it.accountId }, { it.balance })
                    }
                    else -> oldItem == newItem
                }
    }
}
